File: sihati_mobile/lib/core/services/ai_service.dart
// lib/core/services/ai_service.dart
import 'package:get/get.dart';
import '../../app/constants/api_constants.dart';
import 'api_service.dart';

// ─── Models ───────────────────────────────────────────────────

enum UrgencyLevel { low, medium, high, emergency }

UrgencyLevel urgencyFromString(String? s) {
  switch (s) {
    case 'emergency':
      return UrgencyLevel.emergency;
    case 'high':
      return UrgencyLevel.high;
    case 'medium':
      return UrgencyLevel.medium;
    default:
      return UrgencyLevel.low;
  }
}

class ChatResponse {
  final String reply;
  final UrgencyLevel urgency;
  final bool isSymptomRelated;
  final String? suggestedSpecialty;
  final List<AIMedicationResult> medicationSuggestions;

  ChatResponse.fromJson(Map<String, dynamic> j)
      : reply = j['reply'] ?? '',
        urgency = urgencyFromString(j['urgency']),
        isSymptomRelated = j['isSymptomRelated'] ?? false,
        suggestedSpecialty = j['suggestedSpecialty'],
        medicationSuggestions = (j['medicationSuggestions'] as List? ?? [])
            .map((m) => AIMedicationResult.fromJson(m))
            .toList();
}

class AIMedicationResult {
  final String name;
  final String? genericName;
  final String? category;
  final bool requiresPrescription;
  final double? basePrice;
  final bool foundInDb;
  final List<PharmacyStock> availableInPharmacies;

  AIMedicationResult.fromJson(Map<String, dynamic> j)
      : name = j['name'] ?? '',
        genericName = j['genericName'],
        category = j['category'],
        requiresPrescription = j['requiresPrescription'] ?? false,
        basePrice = (j['basePrice'] as num?)?.toDouble(),
        foundInDb = j['foundInDb'] ?? false,
        availableInPharmacies = (j['availableInPharmacies'] as List? ?? [])
            .map((p) => PharmacyStock.fromJson(p))
            .toList();

  bool get hasStock => availableInPharmacies.any((p) => p.inStock);
  int get stockCount => availableInPharmacies.where((p) => p.inStock).length;
}

class PharmacyStock {
  final int pharmacyId;
  final String pharmacyName;
  final String wilaya;
  final String phone;
  final bool isOnDutyTonight;
  final bool inStock;
  final double? price;
  final double? distance;

  PharmacyStock.fromJson(Map<String, dynamic> j)
      : pharmacyId = j['pharmacyId'] ?? 0,
        pharmacyName = j['pharmacyName'] ?? '',
        wilaya = j['wilaya'] ?? '',
        phone = j['phone'] ?? '',
        isOnDutyTonight = j['isOnDutyTonight'] ?? false,
        inStock = j['inStock'] ?? false,
        price = (j['price'] as num?)?.toDouble(),
        distance = (j['distance'] as num?)?.toDouble();
}

class HistoryItem {
  final String role;
  final String text;

  HistoryItem({required this.role, required this.text});

  Map<String, dynamic> toJson() => {
        'role': role,
        'parts': [
          {'text': text}
        ],
      };
}

// ─── AI Service ───────────────────────────────────────────────

class AIService extends GetxService {
  final ApiService _apiService = Get.find<ApiService>();

  Future<AIService> init() async {
    print('🤖 AI Service initialized — ${ApiConstants.BASE_URL}');
    return this;
  }

  // ─── POST /ai/chat ─────────────────────────────────────
  Future<ChatResponse> sendQuery(
    String message,
    List<HistoryItem> history, {
    double? lat,
    double? lng,
  }) async {
    try {
      final body = <String, dynamic>{
        'message': message,
        'history': history.map((h) => h.toJson()).toList(),
        if (lat != null && lng != null) 'location': {'lat': lat, 'lng': lng},
      };
      final response = await _apiService.post('/ai/chat', data: body);
      final data = response.data['data'] ?? response.data;
      return ChatResponse.fromJson(data);
    } catch (e) {
      print('❌ sendQuery error: $e');
      return ChatResponse.fromJson({
        'reply': _errorMessage(e),
        'urgency': 'low',
        'isSymptomRelated': false,
        'suggestedSpecialty': null,
        'medicationSuggestions': [],
      });
    }
  }

  // ─── POST /ai/interaction ──────────────────────────────
  Future<String> checkDrugInteraction({
    required String medication1,
    required String medication2,
  }) async {
    try {
      final response = await _apiService.post(
        '/ai/interaction',
        data: {'med1': medication1, 'med2': medication2},
      );
      final data = response.data['data'] ?? response.data;
      final safe = data['safe'] ?? true;
      final severity = data['severity'] ?? 'unknown';
      final reply = data['reply'] ?? '';

      if (safe) {
        return '✅ Aucune interaction connue entre $medication1 et $medication2.\n\n'
            'Consultez toujours votre médecin ou pharmacien avant de combiner des médicaments.';
      }

      final String emoji;
      switch (severity) {
        case 'high':
          emoji = '🚫';
          break;
        case 'moderate':
          emoji = '⚠️';
          break;
        default:
          emoji = '⚡';
      }

      return '$emoji Interaction détectée ($severity)\n\n'
          '$reply\n\n'
          '⚠️ Consultez votre médecin ou pharmacien.';
    } catch (e) {
      print('❌ checkDrugInteraction error: $e');
      return "Impossible de vérifier l'interaction. Consultez un pharmacien.";
    }
  }

  // ─── POST /ai/ask-medication ────────────────────────────
  Future<String> askMedicationQuestion({
    required String medicationName,
    required String question,
  }) async {
    try {
      final response = await _apiService.post(
        '/ai/ask-medication',
        data: {
          'medicationName': medicationName,
          'question': question,
        },
      );
      final data = response.data['data'] ?? response.data;
      return data['answer'] as String? ??
          'Désolé, je ne peux pas répondre à cette question.';
    } catch (e) {
      print('❌ askMedicationQuestion error: $e');
      return 'Désolé, une erreur est survenue. Veuillez réessayer.';
    }
  }

  // ─── POST /ai/medication-info ──────────────────────────
  Future<Map<String, dynamic>> getMedicationInfo(String name) async {
    try {
      final response = await _apiService.post(
        '/ai/medication-info',
        data: {'medication': name},
      );
      final data = response.data['data'] ?? response.data;
      return {
        'reply': data['reply'] ?? '',
        'usage': data['usage'] ?? '',
        'dosage': data['dosage'] ?? '',
        'warnings': data['warnings'] ?? '',
        'foundInDb': data['foundInDb'] ?? false,
        'contraindications': data['contraindications'] ?? '',
        'sideEffects': data['sideEffects'] ?? '',
        'pregnancy': data['pregnancy'] ?? '',
        'interactions': data['interactions'] ?? '',
      };
    } catch (e) {
      print('❌ getMedicationInfo error: $e');
      return {
        'reply': 'Impossible de récupérer les informations.',
        'usage': '',
        'dosage': '',
        'warnings': '',
        'foundInDb': false,
        'contraindications': '',
        'sideEffects': '',
        'pregnancy': '',
        'interactions': '',
      };
    }
  }

  // ─── POST /ai/specialty ────────────────────────────────
  Future<Map<String, dynamic>> suggestSpecialty(String symptoms) async {
    try {
      final response = await _apiService.post(
        '/ai/specialty',
        data: {'symptoms': symptoms},
      );
      final data = response.data['data'] ?? response.data;
      return {
        'specialty': data['specialty'] ?? 'Médecine Générale',
        'reason': data['reason'] ?? '',
        'urgency': data['urgency'] ?? 'low',
      };
    } catch (e) {
      print('❌ suggestSpecialty error: $e');
      return {
        'specialty': 'Médecine Générale',
        'reason': 'Consultez un médecin généraliste.',
        'urgency': 'low',
      };
    }
  }

  // ─── GET /ai/conversation/:id ────────────────────────
  Future<Map<String, dynamic>?> getConversation(int id) async {
    try {
      final response = await _apiService.get('/ai/conversation/$id');
      final data = response.data['data'] ?? response.data;
      return data['conversation'] as Map<String, dynamic>;
    } catch (e) {
      print('❌ getConversation error: $e');
      return null;
    }
  }

  // ─── GET /ai/history ───────────────────────────────────
  Future<List<Map<String, dynamic>>> getHistory() async {
    try {
      final response = await _apiService.get('/ai/history');
      final data = response.data['data'] ?? response.data;
      return List<Map<String, dynamic>>.from(data['conversations'] ?? []);
    } catch (e) {
      print('❌ getHistory error: $e');
      return [];
    }
  }

  String _errorMessage(dynamic e) {
    final msg = e.toString();
    if (msg.contains('Failed host lookup') ||
        msg.contains('Network is unreachable')) {
      return '⚠️ Impossible de contacter le serveur.\nVérifiez votre connexion.';
    } else if (msg.contains('Connection refused')) {
      return '⚠️ Connexion refusée — vérifiez l\'URL: ${ApiConstants.BASE_URL}';
    } else if (msg.contains('Timeout')) {
      return '⏱️ Le serveur met trop de temps. Veuillez réessayer.';
    }
    return 'Erreur: $msg';
  }
}

File: sihati-backend/src/services/aiService.ts
// src/services/aiService.ts
import model from '../config/gemini';
import { Medication, Pharmacy } from '../models';
import { Op } from 'sequelize';
import {
  ChatHistoryItem,
  ChatResponse,
  InteractionResponse,
  MedicationInfoResponse,
  SpecialtyResponse,
  AIMedicationResult,
  PharmacyStock,
  UrgencyLevel,
} from '../types';

function safeParseJSON<T>(text: string, fallback: T): T {
  try {
    const cleaned = text
      .replace(/^```json\s*/i, '')
      .replace(/^```\s*/i, '')
      .replace(/\s*```$/i, '')
      .trim();
    return JSON.parse(cleaned) as T;
  } catch {
    return fallback;
  }
}

function haversine(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLng = ((lng2 - lng1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLng / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

const CHAT_PROMPT = (message: string) => `
Tu es un assistant santé algérien. Un patient dit: "${message}"

Réponds UNIQUEMENT avec un objet JSON valide (sans markdown, sans backticks):
{
  "reply": "Ta réponse en français (3-5 phrases max, bienveillante et claire)",
  "urgency": "low|medium|high|emergency",
  "isSymptomRelated": true|false,
  "suggestedSpecialty": "Nom exact de la spécialité ou null",
  "medicationNames": ["nom1", "nom2"]
}

Règles urgency:
- emergency: douleur thoracique, difficultés respiratoires, perte de conscience, AVC, hémorragie
- high: fièvre >39°C, douleur intense, symptômes qui s'aggravent rapidement
- medium: symptômes modérés depuis plus de 48h
- low: symptômes légers, question d'information générale

Règles medicationNames:
- Seulement si isSymptomRelated = true
- Maximum 4 médicaments OTC disponibles en Algérie (ex: Doliprane, Spasfon, Aspégic)
- [] si aucun applicable ou si urgency = emergency
`.trim();

const INTERACTION_PROMPT = (med1: string, med2: string) => `
Vérifie l'interaction médicamenteuse entre "${med1}" et "${med2}".
Réponds UNIQUEMENT avec un JSON valide (sans markdown):
{
  "safe": true|false,
  "severity": "none|mild|moderate|severe",
  "reply": "Explication courte en français (2-3 phrases)"
}
`.trim();

const MED_INFO_PROMPT = (name: string, dbContext: string) => `
Tu es un pharmacien expert en Algérie. Donne des informations détaillées sur le médicament "${name}".
${dbContext ? `Données de notre base: ${dbContext}` : ''}
Réponds UNIQUEMENT avec un JSON valide (sans markdown), tous les textes en français:
{
  "reply": "Résumé court et utile pour le patient (2-3 phrases)",
  "usage": "Pour quelles maladies ou symptômes ce médicament est utilisé",
  "contraindications": "Qui ne doit pas prendre ce médicament",
  "dosage": "Dosage adulte standard, fréquence et durée",
  "sideEffects": "Effets secondaires courants et graves",
  "pregnancy": "Sécurité pendant la grossesse et l'allaitement",
  "interactions": "Médicaments à éviter avec celui-ci",
  "warnings": "Précautions importantes, conservation, ordonnance requise oui/non"
}
`.trim();

const SPECIALTY_PROMPT = (symptoms: string) => `
Un patient algérien décrit: "${symptoms}"
Quelle spécialité médicale devrait-il consulter?
Réponds UNIQUEMENT avec un JSON valide (sans markdown):
{
  "specialty": "Nom exact de la spécialité",
  "reason": "Raison courte en français (1 phrase)",
  "urgency": "low|medium|high|emergency"
}
`.trim();

class AIService {

  async sendQuery(
    message: string,
    history: ChatHistoryItem[] = [],
    location?: { lat: number; lng: number }
  ): Promise<ChatResponse> {
    try {
      let trimmedHistory = history.slice(-10);

      while (trimmedHistory.length > 0 && trimmedHistory[0].role !== 'user') {
        trimmedHistory = trimmedHistory.slice(1);
      }

      const chat = model.startChat({ history: trimmedHistory });
      const result = await chat.sendMessage(CHAT_PROMPT(message));
      const raw = result.response.text();

      const parsed = safeParseJSON<any>(raw, null);

      if (!parsed) {
        return {
          reply: raw,
          urgency: 'low',
          isSymptomRelated: false,
          suggestedSpecialty: null,
          medicationSuggestions: [],
        };
      }

      let medicationSuggestions: AIMedicationResult[] = [];
      if (
        parsed.isSymptomRelated &&
        Array.isArray(parsed.medicationNames) &&
        parsed.medicationNames.length > 0
      ) {
        medicationSuggestions = await this._lookupMedications(
          parsed.medicationNames,
          location
        );
      }

      return {
        reply: parsed.reply ?? raw,
        urgency: (parsed.urgency as UrgencyLevel) ?? 'low',
        isSymptomRelated: parsed.isSymptomRelated ?? false,
        suggestedSpecialty: parsed.suggestedSpecialty ?? null,
        medicationSuggestions,
      };
    } catch (error: any) {
      console.error('AIService.sendQuery error:', error);
      const is429 = error?.status === 429 || error?.message?.includes('429');
      return {
        reply: is429
          ? "Le service IA est temporairement surchargé. Veuillez réessayer dans quelques instants."
          : "Désolé, je n'arrive pas à traiter votre demande. Veuillez réessayer.",
        urgency: 'low',
        isSymptomRelated: false,
        suggestedSpecialty: null,
        medicationSuggestions: [],
      };
    }
  }

  private async _lookupMedications(
    names: string[],
    location?: { lat: number; lng: number }
  ): Promise<AIMedicationResult[]> {
    try {
      const nameConditions = names.map((n) => ({
        [Op.or]: [
          { name: { [Op.iLike]: `%${n}%` } },
          { genericName: { [Op.iLike]: `%${n}%` } },
        ],
      }));

      const medications = await Medication.findAll({
        where: { [Op.or]: nameConditions },
        include: [
          {
            model: Pharmacy,
            as: 'pharmacies',
            through: { attributes: ['inStock', 'price'] },
          },
        ],
        limit: 4,
      });

      const results: AIMedicationResult[] = medications.map((med) => {
        const pharmacyRows = ((med as any).pharmacies ?? []).map((p: any): PharmacyStock => {
          const junction = p.PharmacyMedication ?? p.pharmacy_medications;
          const dist = location
            ? Math.round(haversine(location.lat, location.lng, Number(p.latitude), Number(p.longitude)) * 10) / 10
            : undefined;

          return {
            pharmacyId: p.id,
            pharmacyName: p.pharmacyName,
            wilaya: p.wilaya,
            phone: p.phone,
            isOnDutyTonight: p.isOnDutyTonight,
            inStock: junction?.inStock ?? false,
            price: junction?.price ? Number(junction.price) : null,
            distance: dist,
          };
        });

        pharmacyRows.sort((a: PharmacyStock, b: PharmacyStock) => {
          if (a.distance !== undefined && b.distance !== undefined) {
            return a.distance - b.distance;
          }
          return Number(b.isOnDutyTonight) - Number(a.isOnDutyTonight);
        });

        return {
          name: med.name,
          genericName: med.genericName ?? null,
          category: med.category ?? null,
          requiresPrescription: med.requiresPrescription,
          basePrice: null, // Medication model has no price field
          foundInDb: true,
          availableInPharmacies: pharmacyRows.slice(0, 5),
        };
      });

      const foundNames = results.map((r) => r.name.toLowerCase());
      for (const name of names) {
        const alreadyFound = foundNames.some((fn) => fn.includes(name.toLowerCase()));
        if (!alreadyFound) {
          results.push({
            name,
            genericName: null,
            category: null,
            requiresPrescription: false,
            basePrice: null,
            foundInDb: false,
            availableInPharmacies: [],
          });
        }
      }

      return results;
    } catch (error) {
      console.error('AIService._lookupMedications error:', error);
      return [];
    }
  }

  async checkDrugInteraction(med1: string, med2: string): Promise<InteractionResponse> {
    try {
      const result = await model.generateContent(INTERACTION_PROMPT(med1, med2));
      const parsed = safeParseJSON<any>(result.response.text(), null);
      if (!parsed) throw new Error('parse failed');

      return {
        safe: parsed.safe ?? true,
        severity: parsed.severity ?? 'unknown',
        reply: parsed.reply,
      };
    } catch (error) {
      console.error('AIService.checkDrugInteraction error:', error);
      return {
        safe: true,
        severity: 'unknown',
        reply: "Impossible de vérifier l'interaction. Consultez un pharmacien.",
      };
    }
  }

  async getMedicationInfo(medicationName: string): Promise<MedicationInfoResponse> {
    try {
      const dbMed = await Medication.findOne({
        where: {
          [Op.or]: [
            { name: { [Op.iLike]: `%${medicationName}%` } },
            { genericName: { [Op.iLike]: `%${medicationName}%` } },
          ],
        },
      });

      const dbContext = dbMed
        ? `nom=${dbMed.name}, générique=${dbMed.genericName ?? 'N/A'}, forme=${dbMed.form ?? 'N/A'}, dosage=${dbMed.dosage ?? 'N/A'}, catégorie=${dbMed.category ?? 'N/A'}`
        : '';

      const result = await model.generateContent(MED_INFO_PROMPT(medicationName, dbContext));
      const parsed = safeParseJSON<any>(result.response.text(), null);

      return {
        reply:             parsed?.reply ?? result.response.text(),
        usage:             parsed?.usage ?? '',
        contraindications: parsed?.contraindications ?? '',
        dosage:            parsed?.dosage ?? '',
        sideEffects:       parsed?.sideEffects ?? '',
        pregnancy:         parsed?.pregnancy ?? '',
        interactions:      parsed?.interactions ?? '',
        warnings:          parsed?.warnings ?? '',
        foundInDb:         !!dbMed,
        dbData: dbMed
          ? {
              name:                 dbMed.name,
              genericName:          dbMed.genericName ?? null,
              price:                null, // No price column on Medication
              requiresPrescription: dbMed.requiresPrescription,
              category:             dbMed.category ?? null,
              dosageForm:           dbMed.form ?? null,
              strength:             dbMed.dosage ?? null,
            }
          : undefined,
      };
    } catch (error) {
      console.error('AIService.getMedicationInfo error:', error);
      return {
        reply:             'Impossible de récupérer les informations.',
        usage:             '',
        contraindications: '',
        dosage:            '',
        sideEffects:       '',
        pregnancy:         '',
        interactions:      '',
        warnings:          '',
        foundInDb:         false,
      };
    }
  }

  async askMedicationQuestion(medicationName: string, question: string): Promise<{ answer: string }> {
    try {
      const prompt = `
Tu es un pharmacien expert en Algérie. Un patient te pose une question sur le médicament "${medicationName}".

Question: "${question}"

Réponds en français, de façon claire et concise (2-4 phrases max).
Ne donne pas de diagnostic médical. Si la question dépasse tes compétences, recommande de consulter un médecin.
Réponds UNIQUEMENT avec un JSON valide (sans markdown):
{ "answer": "Ta réponse ici" }
`.trim();

      const result = await model.generateContent(prompt);
      const parsed = safeParseJSON<any>(result.response.text(), null);

      return {
        answer: parsed?.answer ?? result.response.text(),
      };
    } catch (error) {
      console.error('AIService.askMedicationQuestion error:', error);
      return {
        answer: 'Désolé, impossible de répondre à cette question. Consultez un pharmacien.',
      };
    }
  }

  async suggestSpecialty(symptoms: string): Promise<SpecialtyResponse> {
    try {
      const result = await model.generateContent(SPECIALTY_PROMPT(symptoms));
      const parsed = safeParseJSON<any>(result.response.text(), null);
      if (!parsed) throw new Error('parse failed');

      return {
        specialty: parsed.specialty ?? 'Médecine Générale',
        reason: parsed.reason ?? '',
        urgency: (parsed.urgency as UrgencyLevel) ?? 'low',
      };
    } catch (error) {
      console.error('AIService.suggestSpecialty error:', error);
      return {
        specialty: 'Médecine Générale',
        reason: 'Consultez un médecin généraliste.',
        urgency: 'low',
      };
    }
  }
}

export default new AIService();